ocaml/xapi/repository.ml

```diff
@@ -28,23 +28,36 @@ let updates_in_cache : (API.ref_host, Yojson.Basic.t) Hashtbl.t =
   Hashtbl.create 64
 
 let introduce ~__context ~name_label ~name_description ~binary_url ~source_url
-    ~update ~gpgkey_path =
-  assert_url_is_valid ~url:binary_url ;
-  assert_url_is_valid ~url:source_url ;
-  assert_gpgkey_path_is_valid gpgkey_path ;
+    ~update ~gpgkey_path ~origin =
+  ( match origin with
+  | `remote ->
+      assert_url_is_valid ~url:binary_url ;
+      assert_url_is_valid ~url:source_url
+  | `bundle ->
+      if binary_url <> "" || source_url <> "" then
+        raise Api_errors.(Server_error (bundle_repo_url_should_be_empty, [])) ;
+      if not update then
+        raise Api_errors.(Server_error (bundle_repo_update_should_be_true, []))
+  ) ;
```

Comment:
This is a bit awkward, as there are two clear cases depending on the value of `origin`, which require different further parameters. The errors are also hard to understand.

This could be avoided by adding a new function `repository.introduce_bundle` that does not have the url and update parameters, and leaving the `repository.introduce` function as it was (without `origin`) for the remote repo case.